Issue 6867

Janpot opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**What kind of change does this PR introduce?**

Bug fix for https://github.com/webpack/webpack/issues/6867
<!-- E.g. a bugfix, feature, refactoring, build related change, etc… -->

**Did you add tests for your changes?**

Yes, always
<!-- Note that we won't merge your changes if you don't add tests -->

**If relevant, link to documentation update:**

<!-- Link PR from webpack/webpack.js.org here, or N/A -->

**Summary**
> Explain the **motivation** for making this change. What existing problem does the pull request solve?

Basically, I tried to compile a project with webpack 4, it didn't work.
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->
https://github.com/webpack/webpack/issues/6867 contains more info on what's going on exactly.

**Does this PR introduce a breaking change?**

Only if you rely on it throwing errors when it shouldn't :)
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

Don't merge yet, I just realized 
```js
export default (class A {

});
```
suffers from the same problem, and some other code probably too.

Kovensky commented on lib/Parser.js:
```diff
@@ -1178,7 +1178,10 @@ class Parser extends Tapable {
+		if (
+			statement.declaration.id &&
+			statement.declaration.type !== "FunctionExpression"
+		) {
```
It's possible that this was an attempt to workaround an old acorn bug where it'd incorrectly parse `export default function () {}` as a function expression.

ooflorent replied:
In acorn 4, `export default function() {` was parsed as a `FunctionExpression`. 
Since acorn 5 it is a `FunctionDeclaration`.

Kovensky commented on test/configCases/parsing/issue-6867/test.js:
```diff
@@ -0,0 +1,3 @@
+export default (function hello () {
+	return 1;
+})
```
I'd feel better if there were tests for each permutation.

```js
// function _declaration_, id is null, in native implementations the function is named "default"
export default function() {}
```

```js
// function _expression_, id is null, in native implementations the function is named "default"
export default (function() {})
```

```js
// function _declaration_, id is "foo"
export default function foo() {
  foo = function() {} // this not working is currently a bug in webpack 😝
})
```

```js
// function _expression_
export default (function bar() {})
```

Kovensky replied:
These same permutations also apply to `class` declarations, but we probably don't need to test them here 🤔

Janpot commented:
@Kovensky @sokra 
```js
export default (class A {

});
```
is breaking too, how do you feel about 
```js
		if (
			statement.declaration.id &&
			!statement.declaration.type.endsWith("Expression")
		)
```
? is that too hacky?

ooflorent commented:
@Janpot yes it is. You could test both node types.

Janpot commented:
@Kovensky @sokra @ooflorent I think I addressed your comments. 

@Kovensky Not sure why you mention `// this not working is currently a bug in webpack 😝`. Do you think it makes sense to address that issue as part of this PR?

Kovensky commented:
Nah, it's definitely a separate issue. Webpack doesn't actually detect when an export is mutable (mutated) or immutable, it just guesses that `function` and `const` are immutable and everything else is mutable.

Janpot commented:
Could it be that the coverage went down, just because the amount of lines increased more than the amount of statements?

sokra commented on test/configCases/parsing/issue-6867/index.js:
```diff
@@ -0,0 +1,57 @@
+it("should compile default export unnamed function declaration", function() {
```
Could you move this test case from `configCases` to `cases`?

Janpot replied:
done

Kovensky commented on test/cases/parsing/issue-6867/index.js:
```diff
@@ -0,0 +1,57 @@
+		});
+});
+
+it("should compile default export named function expression", function() {
```
`class expression`; same for lines 30, 38 and 45

sokra commented:
Thanks
